services/distribution/src/main/java/app/coronawarn/server/services/distribution/runner/S3Distribution.java

```diff
@@ -63,4 +82,39 @@ public void run(ApplicationArguments args) {
       logger.error("Distribution failed.", e);
     }
   }
+
+  /**
+   * Deletes all diagnosis-key files from S3 that are older than retentionDays.
+   *
+   * @param retentionDays the number of days, that files should be retained on S3.
+   */
+  public void applyRetentionPolicy(int retentionDays) throws MinioException, GeneralSecurityException, IOException {
+    List<S3Object> diagnosisKeysObjects = this.objectStoreAccess.getObjectsWithPrefix("version/v1/"
+        + distributionServiceConfig.getApi().getDiagnosisKeysPath() + "/"
+        + distributionServiceConfig.getApi().getCountryPath() + "/"
+        + distributionServiceConfig.getApi().getCountryGermany() + "/"
+        + distributionServiceConfig.getApi().getDatePath() + "/");
+    final String regex = ".*([0-9]{4}-[0-9]{2}-[0-9]{2}).*";
+    final Pattern pattern = Pattern.compile(regex);
+
+    final LocalDate cutOffDate = LocalDate.now(ZoneOffset.UTC).minusDays(retentionDays);
+
+    diagnosisKeysObjects.stream()
+        .filter(diagnosisKeysObject -> {
+          Matcher matcher = pattern.matcher(diagnosisKeysObject.getObjectName());
+          return matcher.matches() && LocalDate.parse(matcher.group(1), DateTimeFormatter.ISO_LOCAL_DATE)
+              .isBefore(cutOffDate);
+        })
+        .map(diagnosisKeysObject -> {
+          try {
+            return objectStoreAccess.deleteObjectsWithPrefix(diagnosisKeysObject.getObjectName());
```

Comment:
We could consider to not make ``deleteObjectsWithPrefix`` return this list but throw an Exception right away. As far as I can see, the result is not used anywhere else (maybe in tests).